README.md
# ymfmidi

ymfmidi is a MIDI player based on the OPL3 emulation core from [ymfm](https://github.com/aaronsgiles/ymfm).

# Support

### Features

* Supports both 4-operator and 2-operator instruments
* Supports some Roland GS, Yamaha XG, and GM Level 2 features (e.g. multiple instrument banks and percussion channels)
* Can emulate multiple chips at once to increase polyphony
* Can output to WAV files
* Can play files containing multiple songs (XMI, MIDI format 2)
* Supported sequence formats:
    * `.hmi`, `.hmp` HMI Sound Operating System
    * `.mid` Standard MIDI files (format 0 or 1)
    * `.mus` DMX sound system / Doom engine
    * `.rmi` Microsoft RIFF MIDI
    * `.xmi` Miles Sound System / Audio Interface Library
* Supported instrument patch formats:
    * `.ad`, `.opl` Miles Sound System / Audio Interface Library
    * `.op2` DMX sound system / Doom engine
    * `.tmb` Apogee Sound System
    * `.wopl` Wohlstand OPL3 editor

More sequence and instrument file formats will probably be supported in the future.


# Usage

ymfmidi can be used as a standalone program (with SDL2), or incorporated into other software to provide easy OPL3-based MIDI playback.

For standalone usage instructions, run the player with no arguments. The player uses patches from [DMXOPL](https://github.com/sneakernets/DMXOPL) by default, but an alternate patch set can be specified as a command-line argument.

To incorporate ymfmidi into your own programs, include everything in the `src` and `ymfm` directories (except for `src/main.cpp` and `src/console.*`), preferably in its own subdirectory somewhere. After that, somewhere in your code:
* `#include "player.h"`
* Create an instance of `OPLPlayer`, optionally specifying a type of chip (the default is `OPLPlayer::ChipOPL3`) and number of chips to emulate (the default is 1)
* Call the `loadSequence` and `loadPatches` methods to load music and instrument data from a path, an existing `FILE*`, or a buffer in memory
* (Optional) Call the `setLoop`, `setSampleRate`, `setGain`, and `setFilter` methods to set up playback parameters
* Periodically call one of the `generate` methods to output audio in either signed 16-bit or floating-point format
* (Optional) Call the `reset` method to restart playback at the beginning

A proper static lib build method will be available sooner or later.

### Real-time MIDI control

In addition to loading a MIDI file, it's also possible to send MIDI messages to an `OPLPlayer` instance in real time using some of its public methods.

To send a standard two- or three-byte MIDI message:

```
	void midiEvent(uint8_t status, uint8_t data0, uint8_t data1 = 0);
```

Helper methods are also available for the most common messages:

```
	void midiNoteOn(uint8_t channel, uint8_t note, uint8_t velocity);
	void midiNoteOff(uint8_t channel, uint8_t note);
	void midiPitchControl(uint8_t channel, double pitch); // range is -1.0 to 1.0
	void midiProgramChange(uint8_t channel, uint8_t patchNum);
	void midiControlChange(uint8_t channel, uint8_t control, uint8_t value);
	
	void midiSetBendRange(uint8_t channel, uint8_t semitones);
```

System Exclusive messages can also be sent directly to the player (primarily for enabling supported GS or XG features):

```
	void midiSysEx(const uint8_t *data, uint32_t length);
```

It is not required to include the opening `0xF0` byte that normally precedes a sysex event; this is due mainly to the way that these events are stored in MIDI files. If `data` includes this opening byte, it should also be included in `length`, but will otherwise be ignored.

# License

ymfmidi and the underlying ymfm library are both released under the 3-clause BSD license.



CHANGES.txt
v0.5.0 [2022-09-28]
- Added option to simulate OPL or OPL2 instead of OPL3
- Added configurable highpass filter to reduce DC offset (default cutoff is 5Hz)
- Fixed calculation of pitch bends, especially for non-default bend ranges
- Improved voice allocation behavior slightly (again)
- .op2 (DMX) patch loader now sanitizes some patches with bogus 2-voice settings

v0.4.0 [2021-05-31]
- Added support for .hmi, .hmp, .rmi files
- Fixed playback of .mus files with invalid control values
- Player app now falls back to 16-bit audio if floating-point isn't supported
- Player app UI handles long file names and instrument names a little better
- OPLPlayer API now supports loading song/patch data from a buffer (using pointer/length pair)

v0.3.0 [2021-05-27]
- Added basic support for GS/XG/GM2 features
- Added support for .xmi files
- Added support for files containing multiple songs
- Added support for RPN #0 (pitch bend range)
- More tweaks to voice/channel allocation

v0.2.0 [2021-05-23]
- Added support for 4-operator instrument patches
- Added support for .ad/.opl (Miles/AIL) and .tmb (Apogee) bank formats
- Slight improvements to voice/channel management
- Fixed clamping of 16-bit output samples when using multiple chips
- Fixed possible issue w/ playback not starting correctly without being manually restarted
  
v0.1.0 [2021-05-20]
- Initial release


src/main.cpp
#include <cstdio>
#include <getopt.h>
#include <signal.h>
#include <unistd.h>

#define SDL_MAIN_HANDLED
extern "C" {
#include <SDL2/SDL.h>
}

#include "console.h"
#include "player.h"

#define VERSION "0.5.0"

static bool g_running = true;
static bool g_paused = false;
static bool g_looping = true;

static void mainLoopSDL(OPLPlayer* player, int bufferSize, bool interactive);
static void mainLoopWAV(OPLPlayer* player, const char* path);

// ----------------------------------------------------------------------------
void usage()
{
	fprintf(stderr, 
	"usage: ymfmidi [options] song_path [patch_path]\n"
	"\n"
	"supported song formats:  HMI, HMP, MID, MUS, RMI, XMI\n"
	"supported patch formats: AD, OPL, OP2, TMB, WOPL\n"
	"\n"
	"supported options:\n"
	"  -h / --help             show this information and exit\n"
	"  -q / --quiet            quiet (run non-interactively)\n"
	"  -1 / --play-once        play only once and then exit\n"
	"  -s / --song <num>       select an individual song, if multiple in file\n"
	"                            (default 1)\n"
	"  -o / --out <path>       output to WAV file (implies -q and -1)\n"
	"\n"
	"  -c / --chip <num>       set type of chip (1 = OPL, 2 = OPL2, 3 = OPL3; default 3)\n"
	"  -n / --num <num>        set number of chips (default 1)\n"
	"  -b / --buf <num>        set buffer size (default 4096)\n"
	"  -g / --gain <num>       set gain amount (default 1.0)\n"
	"  -r / --rate <num>       set sample rate (default 44100)\n"
	"  -f / --filter <num>     set highpass cutoff in Hz (default 5.0)\n"
	"\n"
	);
	
	exit(1);
}

static const option options[] = 
{
	{"help",      0, nullptr, 'h'},
	{"quiet",     0, nullptr, 'q'},
	{"play-once", 0, nullptr, '1'}, 
	{"song",      1, nullptr, 's'}, 
	{"chip",      1, nullptr, 'c'},
	{"num",       1, nullptr, 'n'},
	{"buf",       1, nullptr, 'b'},
	{"gain",      1, nullptr, 'g'},
	{"rate",      1, nullptr, 'r'},
	{"filter",    1, nullptr, 'f'},
	{0}
};

// ----------------------------------------------------------------------------
void quit(int)
{
	g_running = false;
}

// ----------------------------------------------------------------------------
const char* shortPath(const char* path)
{
	const char* p;
	if ((p = strrchr(path, '\\'))
	    || (p = strrchr(path, '/')))
		return p + 1;
	
	return path;
}

// ----------------------------------------------------------------------------
int main(int argc, char **argv)
{
	setbuf(stdout, NULL);
	
	bool interactive = true;
	
	const char* songPath;
	const char* patchPath = "GENMIDI.wopl";
	const char* wavPath = nullptr;
	int sampleRate = 44100;
	int bufferSize = 4096;
	double gain = 1.0;
	double filter = 5.0;
	OPLPlayer::ChipType chipType = OPLPlayer::ChipOPL3;
	int numChips = 1;
	unsigned songNum = 0;

	printf("ymfmidi v" VERSION " - " __DATE__ "\n");

	char opt;
	while ((opt = getopt_long(argc, argv, ":hq1s:o:c:n:b:g:r:f:", options, nullptr)) != -1)
	{
		switch (opt)
		{
		case ':':
		case 'h':
			usage();
			break;
		
		case 'q':
			interactive = false;
			break;
		
		case '1':
			g_looping = false;
			break;
		
		case 's':
			songNum = atoi(optarg);
			break;
		
		case 'o':
			wavPath = optarg;
			interactive = g_looping = false;
			break;
		
		case 'c':
			switch (atoi(optarg))
			{
			case 1: chipType = OPLPlayer::ChipOPL; break;
			case 2: chipType = OPLPlayer::ChipOPL2; break;
			case 3: chipType = OPLPlayer::ChipOPL3; break;
			default:
				fprintf(stderr, "invalid chip type\n");
				exit(1);
			}
			break;
		
		case 'n':
			numChips = atoi(optarg);
			if (numChips < 1)
			{
				fprintf(stderr, "number of chips must be at least 1\n");
				exit(1);
			}
			break;
		
		case 'b':
			bufferSize = atoi(optarg);
			if (!bufferSize)
			{
				fprintf(stderr, "invalid buffer size: %s\n", optarg);
				exit(1);
			}
			break;
		
		case 'g':
			gain = atof(optarg);
			if (!gain)
			{
				fprintf(stderr, "invalid gain: %s\n", optarg);
				exit(1);
			}
			break;
		
		case 'r':
			sampleRate = atoi(optarg);
			if (!sampleRate)
			{
				fprintf(stderr, "invalid sample rate: %s\n", optarg);
				exit(1);
			}
			break;
		
		case 'f':
			filter = atof(optarg);
			if (filter < 0.0)
			{
				fprintf(stderr, "invalid cutoff: %s\n", optarg);
				exit(1);
			}
			break;
		}
	}
	
	if (optind >= argc)
		usage();
	
	songPath = argv[optind];
	if (optind + 1 < argc)
		patchPath = argv[optind + 1];
	
	auto player = new OPLPlayer(numChips, chipType);
	
	if (!player->loadSequence(songPath))
	{
		fprintf(stderr, "couldn't load %s\n", songPath);
		exit(1);
	}
	
	if (!player->loadPatches(patchPath))
	{
		fprintf(stderr, "couldn't load %s\n", patchPath);
		exit(1);
	}
	
	player->setLoop(g_looping);
	player->setSampleRate(sampleRate);
	player->setGain(gain);
	player->setFilter(filter);
	if (songNum > 0)
		player->setSongNum(songNum - 1);
	
	if (interactive)
	{
		consoleOpen();
		consolePos(0);
		printf("song: %-32.32s | patches: %-29.29s\n",
			shortPath(songPath), shortPath(patchPath));
	}
	else
	{
		printf("song:    %s\npatches: %s\n",
			shortPath(songPath), shortPath(patchPath));
	}

	signal(SIGINT, quit);

	if (wavPath)
		mainLoopWAV(player, wavPath);
	else
		mainLoopSDL(player, bufferSize, interactive);
	
	delete player;
	
	return 0;
}

// ----------------------------------------------------------------------------
static SDL_AudioSpec g_audioSpec;

static void audioCallback(void *data, uint8_t *stream, int len)
{
	memset(stream, g_audioSpec.silence, len);
	
	auto player = reinterpret_cast<OPLPlayer*>(data);
	if (g_audioSpec.format == AUDIO_F32SYS)
		player->generate(reinterpret_cast<float*>(stream), len / (2 * sizeof(float)));
	else if (g_audioSpec.format == AUDIO_S16SYS)
		player->generate(reinterpret_cast<int16_t*>(stream), len / (2 * sizeof(int16_t)));
	
	if (!g_looping)
		g_running &= !player->atEnd();
}

// ----------------------------------------------------------------------------
static void mainLoopSDL(OPLPlayer *player, int bufferSize, bool interactive)
{
	// init SDL audio now
	SDL_SetMainReady();
	SDL_Init(SDL_INIT_AUDIO);
	
	SDL_AudioSpec spec = {0};
	spec.freq     = player->sampleRate();
	spec.format   = AUDIO_F32SYS;
	spec.channels = 2;
	spec.samples  = bufferSize;
	spec.callback = audioCallback;
	spec.userdata = player;
	
	if (SDL_OpenAudio(&spec, &g_audioSpec))
	{
		fprintf(stderr, "couldn't open audio device\n");
		exit(1);
	}
	else if (g_audioSpec.format != AUDIO_F32SYS && g_audioSpec.format != AUDIO_S16SYS)
	{
		fprintf(stderr, "unsupported audio format (0x%x)\n", g_audioSpec.format);
		exit(1);
	}
	
	player->setSampleRate(g_audioSpec.freq);
	
	if (interactive)
	{
		consolePos(2);
		printf("\ncontrols: [p] pause, [r] restart, [tab] change view, [esc/q] quit\n");
	}

	SDL_PauseAudio(0);
	
	unsigned displayType = 0;
	while (g_running)
	{
		if (interactive)
		{
			if (player->numSongs() > 1)
			{
				consolePos(1);
				printf("part %3u/%-3u (use left/right to change)\n",
					player->songNum() + 1, player->numSongs());
			}
			
			consolePos(5);
			if (!displayType)
				player->displayChannels();
			else
				player->displayVoices();
			
			switch (consoleGetKey())
			{
			case 0x1b:
			case 'q':
				g_running = false;
				SDL_PauseAudio(1);
				continue;
			
			case 'p':
				g_paused ^= true;
				SDL_PauseAudio(g_paused);
				break;
			
			case 'r':
				g_paused = false;
				SDL_PauseAudio(0);
				player->reset();
				break;
				
			case 0x09:
				displayType ^= 1;
				consolePos(5);
				player->displayClear();
				break;
			
			case -'D':
				if (player->songNum() > 0)
					player->setSongNum(player->songNum() - 1);
				break;

			case -'C':
				if (player->songNum() < player->numSongs() - 1)
					player->setSongNum(player->songNum() + 1);
				break;
			}
		}
		SDL_Delay(30);
	}

	SDL_Quit();
}

// ----------------------------------------------------------------------------
static void mainLoopWAV(OPLPlayer *player, const char *path)
{
	FILE *wav = fopen(path, "wb");
	if (!wav)
	{
		fprintf(stderr, "couldn't open %s\n", path);
		exit(1);
	}
	
	printf("rendering %s...\n", path);
	
	fseek(wav, 44, SEEK_SET);
	
	uint32_t numSamples = 0;
	uint16_t samples[2];
	char outSamples[4];
	const unsigned bytesPerSample = ((player->chipType() == OPLPlayer::ChipOPL3) ? 4 : 2);
	
	while (!player->atEnd())
	{
		player->generate(reinterpret_cast<int16_t*>(samples), 1);
		outSamples[0] = samples[0];
		outSamples[1] = samples[0] >> 8;
		outSamples[2] = samples[1];
		outSamples[3] = samples[1] >> 8;
		
		if (fwrite(outSamples, 1, bytesPerSample, wav) != bytesPerSample)
		{
			fprintf(stderr, "writing WAV data failed\n");
			exit(1);
		}
		numSamples++;
	}
	
	// fill in the rendered sample size and write the header
	const uint32_t sampleRate = player->sampleRate();
	const uint32_t byteRate = sampleRate * bytesPerSample;
	const uint32_t dataSize = numSamples * bytesPerSample;
	const uint32_t wavSize = dataSize + 36;
	
	char header[44] = {0};
	
	header[0] = 'R';
	header[1] = 'I';
	header[2] = 'F';
	header[3] = 'F';
	header[4] = (char)(wavSize);
	header[5] = (char)(wavSize >> 8);
	header[6] = (char)(wavSize >> 16);
	header[7] = (char)(wavSize >> 24);
	header[8]  = 'W';
	header[9]  = 'A';
	header[10] = 'V';
	header[11] = 'E';
	
	// format chunk
	header[12] = 'f';
	header[13] = 'm';
	header[14] = 't';
	header[15] = ' ';
	header[16] = 16; // chunk size
	header[20] = 1;  // sample format (PCM)
	if (bytesPerSample == 4)
		header[22] = 2;  // stereo
	else
		header[22] = 1;  // mono
	header[24] = (char)(sampleRate);
	header[25] = (char)(sampleRate >> 8);
	header[26] = (char)(sampleRate >> 16);
	header[27] = (char)(sampleRate >> 24);
	header[28] = (char)(byteRate);
	header[29] = (char)(byteRate >> 8);
	header[30] = (char)(byteRate >> 16);
	header[31] = (char)(byteRate >> 24);
	header[32] = 4;  // bytes per sample
	header[34] = 16; // bits per sample
	
	// data chunk
	header[36] = 'd';
	header[37] = 'a';
	header[38] = 't';
	header[39] = 'a';
	header[40] = (char)(dataSize);
	header[41] = (char)(dataSize >> 8);
	header[42] = (char)(dataSize >> 16);
	header[43] = (char)(dataSize >> 24);
	
	fseek(wav, 0, SEEK_SET);
	if (fwrite(header, 1, sizeof(header), wav) != sizeof(header))
	{
		fprintf(stderr, "writing WAV header failed\n");
		exit(1);
	}
	
	fclose(wav);
}
